gateway/src/service.rs

```diff
@@ -0,0 +1,522 @@
+use crate::{EndState, ErrorKind, Refresh, Service, State};
+
+impl From<SqlxError> for Error {
+    fn from(err: SqlxError) -> Self {
+        debug!("internal SQLx error: {err}");
+        Self::source(ErrorKind::Internal, err)
+    }
+}
+
+#[derive(Debug)]
+pub struct Work<W = Project> {
+    project_name: ProjectName,
+    account_name: AccountName,
+    work: W,
+}
+
+#[async_trait]
+impl<'c, W> State<'c> for Work<W>
+where
+    W: State<'c>,
+{
+    type Next = Work<W::Next>;
+
+    type Error = W::Error;
+
+    async fn next<C: Context<'c>>(self, ctx: &C) -> Result<Self::Next, Self::Error> {
+        Ok(Work::<W::Next> {
+            project_name: self.project_name,
+            account_name: self.account_name,
+            work: self.work.next(ctx).await?,
+        })
+    }
+}
+
+impl<'c, W> EndState<'c> for Work<W>
+where
+    W: EndState<'c>,
+{
+    fn is_done(&self) -> bool {
+        self.work.is_done()
+    }
+}
+
+pub struct Worker<Svc, W> {
+    service: Svc,
+    send: Sender<W>,
+    recv: Receiver<W>,
+}
+
+impl<Svc, W> Worker<Svc, W>
+where
+    W: Send,
+{
+    pub fn new(service: Svc) -> Self {
+        let (send, recv) = channel(256);
+        Self {
+            service,
+            send,
+            recv,
+        }
+    }
+}
+
+impl<Svc, W> Worker<Svc, W> {
+    pub fn sender(&self) -> Sender<W> {
+        self.send.clone()
+    }
+}
+
+impl<Svc, W> Worker<Svc, W>
+where
+    Svc: for<'c> Service<'c, State = W, Error = Error>,
+    W: Debug + Send + for<'c> EndState<'c>,
+{
+    pub async fn start(mut self) -> Result<(), Error> {
+        while let Some(work) = self.recv.recv().await {
+            let context = self.service.context();
+
+            // Safety: EndState's transitions are Infallible
+            let next = work.next(&context).await.unwrap();
+
+            match self.service.update(&next).await {
+                Ok(_) => {}
+                Err(err) => info!("failed to update a state: {}\nstate: {:?}", err, next),
+            };
+
+            if !next.is_done() {
+                // Safety: Should only panic if `self.recv` has
+                // hanged, which at this point can't happen
+                self.send.send(next).await.unwrap();
+            }
+        }
+        Ok(())
+    }
+}
+
+pub struct GatewayService {
+    docker: Docker,
+    hyper: HyperClient<HttpConnector, Body>,
+    db: SqlitePool,
+    sender: Mutex<Option<Sender<Work>>>,
+    args: Args,
+}
+
+const DB_PATH: &'static str = "gateway.sqlite";
+
+use crate::auth::User;
+use crate::{auth::Key, AccountName};
+
+impl GatewayService {
+    /// Initialize `GatewayService` and its required dependencies.
+    ///
+    /// * `args` - The [`Args`] with which the service was
+    /// started. Will be passed as [`Context`] to workers and state.
+    pub async fn init(args: Args) -> Arc<Self> {
+        info!("docker host: tcp://127.0.0.1:2735");
+        let docker = Docker::connect_with_http_defaults().unwrap();
```

Comment:
```suggestion
        let docker = Docker::connect_with_local_defaults().unwrap();
```
Just to make it work on Unix and Windows